utils/install.py

```diff
@@ -0,0 +1,1632 @@
+
+    download_url = urllib.request.urlretrieve
+
+    def reraise(tp, value=None, tb=None):
+        if value is None:
+            value = tp
+        if value.__traceback__ is not tb:
+            raise value.with_traceback(tb)
+        raise value
+
+else:
+    exec("def reraise(tp, value=None, tb=None):\n    raise tp, value, tb\n")
+
+    import urllib
+
+    download_url = urllib.urlretrieve
+
+
+def show_progress(block_num, block_size, total_size):
+    downloaded = block_num * block_size
+
+    print(
+        end=(
+            "\r|%-60s|" % ("=" * int(60 * downloaded / (total_size)))
+            + "%6.2f %%" % (downloaded / (total_size) * 100)
+        )
+    )
+
+    if downloaded < total_size:
+        pass
+    else:
+        print("Downloaded")
+
+
+@contextmanager
+def opened_w_error(filename, mode="r"):
+    try:
+        f = open(filename, mode)
+    except IOError as err:
+        logging.critical("Error opening file: %s error: %s", filename, err.strerror)
+        yield None
+    else:
+        try:
+            yield f
+        finally:
+            f.close()
+
+
+def _is_tarxz(filename):
+    return filename.endswith(".tar.xz")
+
+
+def _is_tar(filename):
+    return filename.endswith(".tar")
+
+
+def _is_targz(filename):
+    return filename.endswith(".tar.gz")
+
+
+def _is_tgz(filename):
+    return filename.endswith(".tgz")
+
+
+def _is_zip(filename):
+    return filename.endswith(".zip")
+
+
+def extract_archive(
+    from_path, ipath, to_path=None, remove_finished=False, env_file_path=None
+):
+    files_extracted = []
+
+    if to_path is None:
+        to_path = dirname(from_path)
+
+    if _is_tar(from_path):
+        with tarfile.open(from_path, "r") as tar:
+            tar.extractall(path=to_path)
+            files_extracted = tar.getnames()
+    elif _is_targz(from_path) or _is_tgz(from_path):
+        with tarfile.open(from_path, "r:gz") as tar:
+            tar.extractall(path=to_path)
+            files_extracted = tar.getnames()
+    elif _is_tarxz(from_path):
+        with tarfile.open(from_path, "r:xz") as tar:
+            tar.extractall(path=to_path)
+            files_extracted = tar.getnames()
+    elif _is_zip(from_path):
+        with zipfile.ZipFile(from_path, "r") as z:
+            z.extractall(to_path)
+            files_extracted = z.namelist()
+    else:
+        reraise(ValueError("Extraction of {} not supported".format(from_path)))
+
+    logging.debug("Writing installed files to %s file", env_file_path)
+    with opened_w_error(env_file_path, "a") as env_file:
+        if env_file is not None:
+            for filename in files_extracted:
+                fname = filename.replace(CONST_ipkg, ipath)
+
+                # Skip if it ends with "wasmedge" as it is going to be removed at a later stage
+                if fname.endswith("wasmedge") and not fname.endswith("bin/wasmedge"):
+                    continue
+
+                # replace wasmedge folder name with include
```

Comment:
I am a little bit confused here. Does this try to replace `lib64` with `lib`? Instead, checking the `lib64` or `lib` first and then using the detected one is better. There are some OSs that only provide `lib64` or `lib.`

Reply:
These lines just don't append any name that ends with wasmedge other than wasmedge binary.
Example: It will not append `$IPATH/lib/wasmedge`, `$IPATH/include/wasmedge` as there is no need to append folder names. However it will append `$IPATH/bin/wasmedge`.

Reply:
Ah-ha, got it.